File: src/test/java/de/fraunhofer/aisec/crymlin/FSMTest.java

package de.fraunhofer.aisec.crymlin;

import static org.junit.jupiter.api.Assertions.*;

import de.fraunhofer.aisec.mark.XtextParser;
import de.fraunhofer.aisec.mark.markDsl.*;
import de.fraunhofer.aisec.mark.markDsl.impl.MarkDslFactoryImpl;
import de.fraunhofer.aisec.markmodel.MRule;
import de.fraunhofer.aisec.markmodel.Mark;
import de.fraunhofer.aisec.markmodel.MarkModelLoader;
import de.fraunhofer.aisec.markmodel.fsm.FSM;
import de.fraunhofer.aisec.markmodel.fsm.Node;
import java.io.File;
import java.net.URL;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Optional;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

class FSMTest {

	private static Mark mark;

	@BeforeAll
	static void startup() throws Exception {

		URL resource = MarkLoadOutputTest.class.getClassLoader().getResource("mark/PoC_MS1/Botan_CipherMode.mark");
		assertNotNull(resource);
		File markPoC1 = new File(resource.getFile());
		assertNotNull(markPoC1);

		XtextParser parser = new XtextParser();
		parser.addMarkFile(markPoC1);

		HashMap<String, MarkModel> markModels = parser.parse();
		mark = new MarkModelLoader().load(markModels, null);
		assertNotNull(mark);
	}

	@Test
	void parseTest() {
		assertEquals(5, mark.getRules().size()); // 5 total
		assertEquals(3, mark.getRules().stream().filter(rule -> rule.getStatement() != null && rule.getStatement().getEnsure() != null
				&& rule.getStatement().getEnsure().getExp() instanceof OrderExpression).count()); // 3 order

		assertEquals(1, mark.getRules().stream().filter(x -> x.getName().equals("BlockCiphers")).count());
		assertEquals(1, mark.getRules().stream().filter(x -> x.getName().equals("UseOfBotan_CipherMode")).count());
		assertEquals(1, mark.getRules().stream().filter(x -> x.getName().equals("SimpleUseOfBotan_CipherMode")).count());
		assertEquals(1, mark.getRules().stream().filter(x -> x.getName().equals("SimpleUseOfBotan2_CipherMode")).count());
		assertEquals(1, mark.getRules().stream().filter(x -> x.getName().equals("UseRandomIV")).count());
	}

	//  @Test
	//  void fsmTest() {
	//
	//    FSM.clearDB();
	//    for (MRule rule : mark.getRules()) {
	//      if (rule.getStatement() != null
	//          && rule.getStatement().getEnsure() != null
	//          && rule.getStatement().getEnsure().getExp() instanceof OrderExpression) {
	//        OrderExpression inner = (OrderExpression) rule.getStatement().getEnsure().getExp();
	//        FSM fsm = new FSM();
	//        fsm.sequenceToFSM(inner.getExp());
	//        fsm.pushToDB();
	//      }
	//    }
	//  }

	private FSM load(String ruleName) {
		Optional<MRule> opt = mark.getRules().stream().filter(x -> x.getName().equals(ruleName)).findFirst();
		assertTrue(opt.isPresent());
		MRule rule = opt.get();
		assertNotNull(rule.getStatement());
		assertNotNull(rule.getStatement().getEnsure());
		assertNotNull(rule.getStatement().getEnsure().getExp());
		assertTrue(rule.getStatement().getEnsure().getExp() instanceof OrderExpression);
		OrderExpression inner = (OrderExpression) rule.getStatement().getEnsure().getExp();
		FSM fsm = new FSM();
		assertNotNull(fsm);
		fsm.sequenceToFSM(inner.getExp());

		return fsm;
	}

	@Test
	void testUseOfBotan_CipherMode() {
		FSM fsm = load("UseOfBotan_CipherMode");

		assertEquals("cm.create (0)\n" + "\t-> cm.init(1)\n" + "cm.init (1)\n" + "\t-> cm.start(2)\n" + "cm.start (2)\n" + "\t-> cm.finish(3)\n" + "\t-> cm.process(4)\n"
				+ "cm.finish (3)\n" + "\t-> END(5)\n" + "\t-> cm.reset(6)\n" + "\t-> cm.start(2)\n" + "cm.process (4)\n" + "\t-> cm.finish(3)\n" + "\t-> cm.process(4)\n"
				+ "END (5)\n" + "cm.reset (6)\n" + "\t-> END(5)\n",
			fsm.toString());
	}

	@Test
	void testSimpleUseOfBotan_CipherMode() {
		FSM fsm = load("SimpleUseOfBotan_CipherMode");

		assertEquals("cm.create (0)\n" + "\t-> cm.init(1)\n" + "cm.init (1)\n" + "\t-> cm.finish(2)\n" + "\t-> cm.start(3)\n" + "cm.finish (2)\n" + "\t-> END(4)\n"
				+ "cm.start (3)\n" + "\t-> cm.finish(2)\n" + "\t-> cm.start(3)\n" + "END (4)\n",
			fsm.toString());
	}

	@Test
	void testSimpleUseOfBotan2_CipherMode() {
		FSM fsm = load("SimpleUseOfBotan2_CipherMode");

		assertEquals("cm.create (0)\n" + "\t-> cm.init(1)\n" + "cm.init (1)\n" + "\t-> cm.start(2)\n" + "cm.start (2)\n" + "\t-> cm.finish(3)\n" + "\t-> cm.start(2)\n"
				+ "cm.finish (3)\n" + "\t-> END(4)\n" + "END (4)\n",
			fsm.toString());
	}

	@Test
	void testRegexToFsm() {
		FSM fsm = new FSM();

		// Create a regular expression: AB?(CD)*E+
		MarkDslFactory f = new MarkDslFactoryImpl();
		Terminal a = f.createTerminal();
		a.setEntity("A");
		Terminal b = f.createTerminal();
		b.setEntity("B");
		Terminal c = f.createTerminal();
		c.setEntity("C");
		Terminal d = f.createTerminal();
		d.setEntity("D");
		Terminal e = f.createTerminal();
		e.setEntity("E");
		RepetitionExpression someB = f.createRepetitionExpression();
		someB.setOp("?");
		someB.setExpr(b);
		SequenceExpression cThenD = f.createSequenceExpression();
		cThenD.setLeft(c);
		cThenD.setRight(d);
		RepetitionExpression anyCThenD = f.createRepetitionExpression();
		anyCThenD.setOp("*");
		anyCThenD.setExpr(cThenD);
		RepetitionExpression someE = f.createRepetitionExpression();
		someE.setOp("+");
		someE.setExpr(e);
		SequenceExpression tail2 = f.createSequenceExpression();
		tail2.setLeft(anyCThenD);
		tail2.setRight(someE);
		SequenceExpression tail1 = f.createSequenceExpression();
		tail1.setLeft(someB);
		tail1.setRight(tail2);
		SequenceExpression aThenTail = f.createSequenceExpression();
		aThenTail.setLeft(a);
		aThenTail.setRight(tail1);

		// Transform regex into NFA (FSM)
		fsm.sequenceToFSM(aThenTail);
		System.out.println(fsm.toString());

		HashSet<Node> start = fsm.getStart();
		assertEquals(1, start.size());
		Node startNode = start.iterator().next();
		HashSet<Node> expectB = startNode.getSuccessors();
		System.out.println(expectB);
	}
}

File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/FSMTest.kt
package de.fraunhofer.aisec.codyze.crymlin

import de.fraunhofer.aisec.codyze.markmodel.MRule
import de.fraunhofer.aisec.codyze.markmodel.Mark
import de.fraunhofer.aisec.codyze.markmodel.MarkModelLoader
import de.fraunhofer.aisec.codyze.markmodel.fsm.FSM
import de.fraunhofer.aisec.mark.XtextParser
import de.fraunhofer.aisec.mark.markDsl.MarkDslFactory
import de.fraunhofer.aisec.mark.markDsl.OrderExpression
import de.fraunhofer.aisec.mark.markDsl.impl.MarkDslFactoryImpl
import java.io.*
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import org.junit.jupiter.api.*

internal class FSMTest {
    @Test
    fun parseTest() {
        assertEquals(5, mark!!.rules.size) // 5 total
        assertEquals(
            3,
            mark!!
                .rules
                .stream()
                .filter { rule: MRule ->
                    rule.statement != null &&
                        rule.statement.ensure != null &&
                        rule.statement.ensure.exp is OrderExpression
                }
                .count()
        ) // 3 order
        assertEquals(
            1,
            mark!!.rules.stream().filter { x: MRule -> x.name == "BlockCiphers" }.count()
        )
        assertEquals(
            1,
            mark!!.rules.stream().filter { x: MRule -> x.name == "UseOfBotan_CipherMode" }.count()
        )
        assertEquals(
            1,
            mark!!
                .rules
                .stream()
                .filter { x: MRule -> x.name == "SimpleUseOfBotan_CipherMode" }
                .count()
        )
        assertEquals(
            1,
            mark!!
                .rules
                .stream()
                .filter { x: MRule -> x.name == "SimpleUseOfBotan2_CipherMode" }
                .count()
        )
        assertEquals(
            1,
            mark!!.rules.stream().filter { x: MRule -> x.name == "UseRandomIV" }.count()
        )
    }

    //  @Test
    //  void fsmTest() {
    //
    //    FSM.clearDB();
    //    for (MRule rule : mark.getRules()) {
    //      if (rule.getStatement() != null
    //          && rule.getStatement().getEnsure() != null
    //          && rule.getStatement().getEnsure().getExp() instanceof OrderExpression) {
    //        OrderExpression inner = (OrderExpression) rule.getStatement().getEnsure().getExp();
    //        FSM fsm = new FSM();
    //        fsm.sequenceToFSM(inner.getExp());
    //        fsm.pushToDB();
    //      }
    //    }
    //  }
    private fun load(ruleName: String): FSM {
        val opt = mark!!.rules.stream().filter { x: MRule -> x.name == ruleName }.findFirst()
        assertTrue(opt.isPresent)

        val rule = opt.get()
        assertNotNull(rule.statement)
        assertNotNull(rule.statement.ensure)
        assertNotNull(rule.statement.ensure.exp)
        assertTrue(rule.statement.ensure.exp is OrderExpression)
        val inner = rule.statement.ensure.exp as OrderExpression
        val fsm = FSM()
        assertNotNull(fsm)

        fsm.sequenceToFSM(inner.exp)
        return fsm
    }

    @Test
    fun testUseOfBotan_CipherMode() {
        val fsm = load("UseOfBotan_CipherMode")
        assertEquals(
            """cm.create (0)
	-> cm.init(1)
cm.init (1)
	-> cm.start(2)
cm.start (2)
	-> cm.finish(3)
	-> cm.process(4)
cm.finish (3)
	-> END (E)(5)
	-> cm.reset(6)
	-> cm.start(2)
cm.process (4)
	-> cm.finish(3)
	-> cm.process(4)
END (E) (5)
cm.reset (6)
	-> END (E)(5)
""",
            fsm.toString()
        )
    }

    @Test
    fun testSimpleUseOfBotan_CipherMode() {
        val fsm = load("SimpleUseOfBotan_CipherMode")
        assertEquals(
            """cm.create (0)
	-> cm.init(1)
cm.init (1)
	-> cm.finish(2)
	-> cm.start(3)
cm.finish (2)
	-> END (E)(4)
cm.start (3)
	-> cm.finish(2)
	-> cm.start(3)
END (E) (4)
""",
            fsm.toString()
        )
    }

    @Test
    fun testSimpleUseOfBotan2_CipherMode() {
        val fsm = load("SimpleUseOfBotan2_CipherMode")
        assertEquals(
            """cm.create (0)
	-> cm.init(1)
cm.init (1)
	-> cm.start(2)
cm.start (2)
	-> cm.finish(3)
	-> cm.start(2)
cm.finish (3)
	-> END (E)(4)
END (E) (4)
""",
            fsm.toString()
        )
    }

    @Test
    fun testRegexToFsm() {
        val fsm = FSM()

        // Create a regular expression: AB?(CD)*E+
        val f: MarkDslFactory = MarkDslFactoryImpl()
        val a = f.createTerminal()
        a.entity = "A"
        val b = f.createTerminal()
        b.entity = "B"
        val c = f.createTerminal()
        c.entity = "C"
        val d = f.createTerminal()
        d.entity = "D"
        val e = f.createTerminal()
        e.entity = "E"
        val someB = f.createRepetitionExpression()
        someB.op = "?"
        someB.expr = b
        val cThenD = f.createSequenceExpression()
        cThenD.left = c
        cThenD.right = d
        val anyCThenD = f.createRepetitionExpression()
        anyCThenD.op = "*"
        anyCThenD.expr = cThenD
        val someE = f.createRepetitionExpression()
        someE.op = "+"
        someE.expr = e
        val tail2 = f.createSequenceExpression()
        tail2.left = anyCThenD
        tail2.right = someE
        val tail1 = f.createSequenceExpression()
        tail1.left = someB
        tail1.right = tail2
        val aThenTail = f.createSequenceExpression()
        aThenTail.left = a
        aThenTail.right = tail1

        // Transform regex into NFA (FSM)
        fsm.sequenceToFSM(aThenTail)
        println(fsm)
        val start = fsm.start
        assertEquals(1, start.size)
        val startNode = start.iterator().next()
        val expectB = startNode.successors
        println(expectB)
    }

    companion object {
        private var mark: Mark? = null
        @BeforeAll
        @JvmStatic
        fun startup() {
            val resource =
                MarkLoadOutputTest::class.java.classLoader.getResource(
                    "mark/PoC_MS1/Botan_CipherMode.mark"
                )
            assertNotNull(resource)
            val markPoC1 = File(resource.file)
            assertNotNull(markPoC1)
            val parser = XtextParser()
            parser.addMarkFile(markPoC1)
            val markModels = parser.parse()
            mark = MarkModelLoader().load(markModels, null)
            assertNotNull(mark)
        }
    }
}
